earthfile2llb/converter.go

```diff
@@ -2416,6 +2428,41 @@ func (c *Converter) internalRun(ctx context.Context, opts ConvertRunOpts) (pllb.
 	}
 }
 
+// detectAWSCredentials determines whether AWS envs are set and/or ~/.aws
+// exists. These are then used when fetching the secrets during a build.
+func (c *Converter) detectAWSCredentials(ctx context.Context) ([]llb.RunOption, []string, error) {
+
+	credsAvail, err := awsutil.CredsAvailable()
+	if err != nil {
+		return nil, nil, errors.Wrap(err, "failed to check for AWS credentials")
+	}
+
+	if !credsAvail {
+		return nil, nil, errors.New("AWS credentials not found in environment or ~/.aws")
+	}
+
+	var (
+		runOpts   = []llb.RunOption{}
+		extraEnvs = []string{}
+	)
+
+	// Add LLB secrets for each of the typical secrets which will then be
+	// sourced from the environment during lookup.
+	for _, secretName := range secretprovider.AWSCredentials {
+		secretPath := path.Join("/run/secrets", secretName)
+		secretOpts := []llb.SecretOption{
+			llb.SecretID(c.secretID(secretName)),
+			llb.SecretFileOpt(0, 0, 0444),
+		}
+		runOpts = append(runOpts, llb.AddSecret(secretPath, secretOpts...))
+		if envName, ok := secretprovider.AWSEnvName(secretName); ok {
+			extraEnvs = append(extraEnvs, fmt.Sprintf("%s=\"$(cat %s)\"", envName, secretPath))
+		}
+	}
+
+	return runOpts, extraEnvs, nil
+}
```

Comment:
Normally the LLB should not be influenced by the local system, because it can cause the same operation to have different cache hashes, causing unexpected cache misses on different systems. LLB should only be influenced by the syntax (e.g. the existence of the `--aws` flag under `RUN`).

Reply:
The local system is only used at this point to determine if there is some sort of support for AWS creds. The code then uses special, internal secret names to help the AWS secret provider expand them based on the local system's env vars or config files. I don't think the LLB will differ depending on the cred storage type.

Reply:
I don't fully understand the logic, but my gut is telling me that something isn't right. LLB should almost never depend on the local system.

Could the logic be made like this:

* Add the LLB only when `--aws` is passed to a `RUN` command.
* If the local system does not have AWS creds to provide, the build results in an error when Buildkit attempts to use the secret provider.

Reply:
Here's how it works at the moment:

1. Check for `--aws` flag on `RUN`
2. If flag exists and creds are available in environment, add internal secret names as LLB
3. Secret values are then loaded into env vars using the new AWS secret provider
4. Secret provider errors when it can't find one of the expected secrets

The LLB only depends on the environment to an extent--in that it checks if the envs or config files exist beforehand. 

Is there a way to provide these secrets without the LLB definitions?

Reply:
@vladaionescu and I spoke about this last night. He suggested removing the upfront checks for credentials to remove potential differences in the generated LLB. The new credential provider will return an error if the requested secret isn't available in the environment.